### ISSUE TITLE
if network af4f49 is finally turned to be a loser(rate<0.5), what shall we do?

### ISSUE THREAD

**Issue by l1t1** [gcp/leela-zero#780]
shall the self-games be removed from the database and train with 7e7106 again?

**Comment by jkiliani** [comment]
Most definitely not. See #764, #778.

**Comment by MaxMaki** [comment]
Maybe the f7ae68fe network could be requeued against the af4f49 one?
It's at 54% against the old best one, so it could perhaps legitimatly dethrone the af4 one?

**Comment by jkiliani** [comment]
It might, or it might not. There's indications match results aren't completely transitive (i.e. if A beats B and B beats C, there's no guarantee that A beats C). In either case, it doesn't matter that much... the training data from af4f49f1 will be useful, and it likely won't take that long to be dethroned anyway.

**Comment by MaxMaki** [comment]
There's a graph in the "how to find information on the best network" thread showing that this network is much stronger in the opening 200 moves or so, and therefor will probably get more early wins.

However, wont this also cause problems for networks to dehtrone it? Getting a false early SPRT fail, just as this network got an early SPRT pass? The problem could work both ways and networks could therefor have a problem beating this if they don't get to play enough matches?

**Comment by jkiliani** [comment]
You'd have a valid point there if an early SPRT fail actually meant that network was already eliminated. However, the server always queues 50 matches initially, so even after an early fail the better results should still come in and make the result null again. At that point, more matches will be queued and the wins from longer matches should eventually tip the balance in favour of the stronger net.

**Comment by MaxMaki** [comment]
Ah, i thought it was 40 games. If i counted correctly this new network went on a 31:9 winning rampage early on. So basically my question is that is the initial batch of games sent out big enough to drag the network out of a SPRT fail, since if i've understood correctly it sends less and less games out if the score looks bad?  If a 31:9 can happen for winning, it can also happen for loosing, and probobly 31:9 isn't even the worst ratio statistically possible.

**Comment by LRGB** [comment]
@roy7 @jkiliani  I do think we could give f7ae68fe another title shot here. It almost promoted against 7e7106d3, which the current best failed to beat. Sure, results are not transitive but I don't see the harm in trying. We've had networks promoted under 55% before, but never under 50% . We are potentially wasting a superior network while training on a subpar one.

**Comment by jkiliani** [comment]
@LRGB Please consider what the Alpha Zero approach would mean for us, and that has been proven by Deepmind to actually work. They promote every new net without testing, and still progress. In our case, that would mean that a whole load of networks that score 20% or less against the current known best would be picked to produce self-play games. Before their paper came out, no-one would have even considered such a self-play pipeline without testing, but it seems to work regardless. This indicates that diversity in play styles and quick feedback may be more important for training progress than raw strength of the networks producing the self-play games. f7ae68fe is nothing more  than a minor hiccup or curiosity in the training process, and it certainly won't hurt the project if it ends up producing games for a while.

**Comment by gcp** [comment]
See issue #764.

We shall do absolutely nothing.

**Comment by wctgit** [comment]
> This indicates that diversity in play styles and quick feedback may be more important for training progress than raw strength of the networks producing the self-play games.

I agree. I've been thinking about ways that a Genetic Algorithm-style free for all tournament between a variety of different nets, with new nets being generated by existing nets every 100 or 200 accumulated wins for example, could potentially help develop a much more robust playing style that not only learns how to play against itself, for example, but how to play against strong nets with a variety of different styles. Kind of like an MMA tournament for AIs. GAs, while typically slow, can be useful at generating robust solutions to problems.

**Comment by jkiliani** [comment]
Turning the training into an application of evolutionary game theory is interesting, but not exactly trivial to apply: The only way to simulate mutation is training of by gradient descent, the same thing we're actually doing already.

I pictured a tournament as a potential approach to diversify training data, but the way I'd do it would be something like this: We keep a "league" of something like 10 reasonably strong nets, and instead of one network playing itself, we do a round-robin tournament of all currently active networks. At the end of a pre-determined time, we score each active network by their winrate against other nets. The weakest net gets kicked out and replaced by a newly trained net at this time, and then the cycle repeats with one net exchanged for a newly trained one.

This would dispense with the evaluation games completely, since self-play in league style would be an evaluation by itself. And in contrast to Alpha Zero, we wouldn't have regression to any large degree, since a weak net will immediately drop out after its first round in the league.

**Comment by wctgit** [comment]
>Turning the training into an application of evolutionary game theory is interesting, but not exactly trivial to apply

I'm not so sure about that. We're already running 'matches' against different networks. The only real difference would be in allowing more than one top network to survive each 'generation', and running matches between randomly selected networks from this population of survivors.

 > The only way to simulate mutation is training of by gradient descent, the same thing we're actually doing already.

Right, but the current implementation only uses the current highest rated network, which fundamentally has the weakness of potentially getting stuck in a local-maximum playing style. By using a 'population' of networks, then alternative playing styles have the opportunity to 'survive' a few generations to see if they can compete better against the 'champ' over time. Also, the 'champ' will be forced to defend not only against similar clones to itself, but to a wider variety of contenders, leading to a more robust 'champ' over time.

> I pictured a tournament as a potential approach to diversify training data, but the way I'd do it would be something like this: ...

That's roughly equivalent to what I'm proposing as well.
